sourmash_lib/__main__.py

```diff
@@ -696,189 +646,190 @@ def categorize(self, args):
-                                                      query_moltype))
+        R_metagenome = 2**64 / float(orig_query.estimator.max_hash)
 
-        tree = SBT.load(args.sbt_name, leaf_loader=SigLeaf.load)
-        #s = sig.load_signatures(args.query, select_ksize=args.ksize)
-        orig_query = query
+        new_mins = query.estimator.get_hashes()
+        e = sourmash_lib.Estimators(ksize=args.ksize, n=len(new_mins))
+        e.update(query.estimator)
+        query = sig.SourmashSignature('', e)
 
         sum_found = 0.
         found = []
         while 1:
-            search_fn = SearchMinHashesFindBest().search
+            search_fn = SearchMinHashesFindBestIgnoreMaxHash().search
 
             results = []
             # use super low threshold for this part of the search
             for leaf in tree.find(search_fn, query, 0.00001):
-                results.append((query.similarity(leaf.data), leaf.data))
-                #results.append((leaf.data.similarity(ss), leaf.data))
+                results.append((query.estimator.similarity_ignore_maxhash(leaf.data.estimator), leaf.data))
 
             if not len(results):          # no matches at all!
                 break
 
             # take the best result
             results.sort(key=lambda x: -x[0])   # reverse sort on similarity
             best_sim, best_ss = results[0]
-            sim = best_ss.similarity(orig_query)
 
-            # adjust by size of leaf (kmer cardinality of original genome)
-            if best_ss.estimator.hll:
-                leaf_kmers = best_ss.estimator.hll.estimate_cardinality()
-                query_kmers = orig_query.estimator.hll.estimate_cardinality()
-                f_of_total = leaf_kmers / query_kmers * sim
+            # subtract found hashes from search hashes, construct new search
+            new_mins = set(query.estimator.get_hashes())
+            found_mins = best_ss.estimator.get_hashes()
+
+            if best_ss.estimator.max_hash:
+                R_genome = 2**64 / float(best_ss.estimator.max_hash)
+            elif best_ss.estimator.hll:
+                genome_size = best_ss.estimator.hll.estimate_cardinality()
+                genome_max_hash = max(found_mins)
+                R_genome = float(genome_size) / float(genome_max_hash)
             else:
-                f_of_total = 0
+                error('Best hash match in sbt_gather has no cardinality')
+                error('Please prepare database of sequences with --scaled')
+                error('...or with --with-cardinality')
+                sys.exit(-1)
 
-            if not found and sim < args.threshold:
-                print('best match: {}'.format(best_ss.name()))
-                print('similarity is {:.5f} of db signature;'.format(sim))
-                print('this is below specified threshold => exiting.')
+            R_comparison = max(R_metagenome, R_genome)
+            new_max_hash = 2**64 / float(R_comparison)
```

Comment:
We don't have any plans on changing hash function size, but maybe move all the `2 ** 64` to a constant in the beginning?